grapher/text/MarkdownTextWrap.tsx

```diff
@@ -0,0 +1,549 @@
+        return <React.Fragment key={key}> </React.Fragment>
+    }
+    toPlaintext(): string {
+        return " "
+    }
+}
+
+export class IRLineBreak implements IRToken {
+    get width(): number {
+        return 0
+    }
+    getBreakpointBefore(): undefined {
+        return undefined
+    }
+    toHTML(key: number | string): JSX.Element {
+        return <br key={key} />
+    }
+    toSVG(): undefined {
+        // We have to deal with this special case in
+        // whatever procedure does text reflow.
+        return undefined
+    }
+    toPlaintext(): string {
+        return "\n"
+    }
+}
+
+export abstract class IRElement implements IRToken {
+    constructor(public children: IRToken[], public fontParams?: IRFontParams) {}
+
+    @imemo get width(): number {
+        return getLineWidth(this.children)
+    }
+
+    getBreakpointBefore(targetWidth: number): IRBreakpoint | undefined {
+        return getBreakpointBefore(this.children, targetWidth)
+    }
+
+    splitBefore(maxWidth: number): {
+        before: IRToken | undefined
+        after: IRToken | undefined
+    } {
+        const { before, after } = splitLineAtBreakpoint(this.children, maxWidth)
+        return {
+            // do not create tokens without children
+            before: before.length ? this.getClone(before) : undefined,
+            after: after.length ? this.getClone(after) : undefined,
+        }
+    }
+
+    splitOnNextLineBreak(): { before?: IRToken; after?: IRToken } {
+        const index = this.children.findIndex(
+            (token) => token instanceof IRLineBreak
+        )
+        if (index >= 0) {
+            return {
+                before:
+                    // do not create an empty element if the first child
+                    // is a newline
+                    index === 0
+                        ? undefined
+                        : this.getClone(this.children.slice(0, index)),
+                after: this.getClone(this.children.slice(index + 1)),
+            }
+        }
+        return { before: this }
+    }
+
+    abstract getClone(children: IRToken[]): IRElement
+    abstract toHTML(key: number | string): JSX.Element
+    abstract toSVG(key: number | string): JSX.Element
+
+    toPlaintext(): string {
+        return lineToPlaintext(this.children)
+    }
+}
+
+export class IRBold extends IRElement {
+    @imemo get width(): number {
+        const plaintext = this.children
+            .map((token) => token.toPlaintext())
+            .join("")
+        return Bounds.forText(plaintext, {
+            ...this.fontParams,
+            fontWeight: 700,
+        }).width
+    }
+
+    getClone(children: IRToken[]): IRBold {
+        return new IRBold(children, this.fontParams)
+    }
+    toHTML(key: number | string): JSX.Element {
+        return (
+            <strong key={key}>
+                {this.children.map((child, i) => child.toHTML(i))}
+            </strong>
+        )
+    }
```

Comment:
nit: could we make `key` optional? (also the `React.Key` type might be slightly better to use, although it just maps to `string | number` so technically the same)

---

Also we could maybe drop the `key` param by wrapping the output of `toHTML` and `toSVG` in `<React.Fragment>`. E.g. there could be a small utility function:

```js
function renderChildren(children, renderFunc) {
    return children.map((child, i) => (
        <React.Fragment key={i}>
            {renderFunc(child)}
        </React.Fragment>
    ))
}
```

and then this becomes

```suggestion
    toHTML(key: number | string): JSX.Element {
        return (
            <strong>
                {renderChildren(children, (child) => child.toHTML())}
            </strong>
        )
    }
```

but not clear that it's better, just makes calling `toHTML()` cleaner.

Reply:
Thanks for the review @danielgavrilov !

Changes:
- `key` is optional now, and typed with `React.Key`
  - I didn't implement `renderChildren`, as it seems like a bit of a leaky abstraction. These classes are probably going to be called via `MarkdownTextWrap` 99% of the time so I don't think it's a huge concern?
- Fixed italic and bold width calculation, including propagating `isItalic` to `pixelWidth`
- Removed `style` - we can wait till we need it
- `whitespace: pre-wrap` fixes the collapsing whitespace problem
  - I stuck with the block span rendering method as it makes calculating line heights easier (text content and br elements behave weirdly with line height when inline, whereas a block span is always just `line-height * font-size`)